addons/stock/models/product_strategy.py

```diff
@@ -63,6 +64,35 @@ def _default_product_id(self):
     package_type_ids = fields.Many2many('stock.package.type', string='Package Type', check_company=True)
     storage_category_id = fields.Many2one('stock.storage.category', 'Storage Category', ondelete='cascade', check_company=True)
     active = fields.Boolean('Active', default=True)
+    sublocation = fields.Selection([
+        ('no', 'No'),
+        ('last_used', 'Last Used'),
+        ('closest_location', 'Closest Location')
+    ], default='no')
+
+    @api.onchange('sublocation', 'location_out_id', 'storage_category_id')
+    def _onchange_sublocation(self):
+        if self.sublocation == 'closest_location':
+            child_location_ids = self.env['stock.location'].search([
+                ('id', 'child_of', self.location_out_id.id),
+                ('storage_category_id', '=', self.storage_category_id.id)
+            ])
+            if not child_location_ids:
+                return {
+                    'warning': {
+                        'title': _("Warning"),
+                        'message': _("Selected storage category does not exist in the 'store to' location or any of its sublocations"),
+                    },
+                }
+        else:
+            self.storage_category_id = None
```

Comment:
avoid using onchange on stuff different than warning. It will introduce a difference between front and backend. Instead use compute (e.g. the storage_category_id could become a compute with a depends on sublocation)

Reply:
`storage_category_id` is not dependent on `sublocation`. However, we need to set `storage_category_id` to `None` when the sublocation is anything but `closest_location`

Reply:
> However, we need to set storage_category_id to None when the sublocation is anything but closest_location

Then it depends on the sublocation no?

Reply:
It can be half-computed. For example it will take the user values but it will be recomputed if the value of sublocation is modified to closest_location. Otherwise it will just keep the current value

Reply:
You can check my fixup. It does the same behavior. It just avoid two different behavior. onchange are call only when coming from client. That way we have the same behavior by modifying the sublocation in client or write putaway.sublocation = ...